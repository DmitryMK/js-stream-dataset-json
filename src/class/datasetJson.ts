import fs from "fs";
import fsPromises from "fs/promises";
import readline from "readline";
import {
    ItemDataArray,
    ItemDataObject,
    DatasetMetadata,
    DataType,
    UniqueValues,
    MetadataAttributes,
    ParsedAttributes,
} from "./../interfaces/datasetJson.d";
import JSONStream from "JSONStream";

// Main class for dataset JSON;
class DatasetJson {
    // Path to the file;
    filePath: string;
    // Statistics about the file;
    stats: fs.Stats;
    // Item Group metadata;
    metadata: DatasetMetadata;

    // Current position in the file;
    currentPosition: number;
    // Flag to indicate if all rows are read;
    allRowsRead: boolean;
    // Metadata loaded
    private metadataLoaded: boolean;
    // Stream
    private stream: fs.ReadStream;
    // Parser
    private isNdJson: boolean;
    // NDJSON stream
    private parser?: fs.ReadStream;
    // NDJSON flag
    private rlStream?: readline.Interface;
    // Required attributes
    private requiredAttributes = [
        "datasetJSONCreationDateTime",
        "datasetJSONVersion",
        "records",
        "name",
        "label",
        "columns",
    ];

    /**
     * Read observations.
     * @constructor
     * @param filePath - Path to the file.
     */
    constructor(filePath: string, options?: { isNdJson?: boolean }) {
        this.filePath = filePath;
        this.currentPosition = 0;
        // If option isNdjson is not specified, try to detect it from the file extension;
        if (options?.isNdJson === undefined) {
            this.isNdJson = this.filePath.toLowerCase().endsWith(".ndjson");
        } else {
            this.isNdJson = options.isNdJson;
        }
        this.allRowsRead = false;
        this.metadataLoaded = false;

        this.metadata = {
            datasetJSONCreationDateTime: "",
            datasetJSONVersion: "",
            records: -1,
            name: "",
            label: "",
            columns: [],
        };

        // Check if file exists;
        if (!fs.existsSync(this.filePath)) {
            throw new Error(`Could not read file ${this.filePath}`);
        }

        this.stats = fs.statSync(this.filePath);

        this.stream = fs.createReadStream(this.filePath, {
            encoding: "utf8",
        });
    }

    /**
     * Check if the file was modified
     * @return True if file has changed, otherwise false.
     */
    private async fileChanged(): Promise<boolean> {
        const stats = await fsPromises.stat(this.filePath);
        if (this.stats !== null && stats.mtimeMs !== this.stats.mtimeMs) {
            return true;
        }
        return false;
    }

    /**
     * Auxilary function to verify if required elements are parsed;
     * @return True if all required attributes are present, otherwise false.
     */
    private checkAttributesParsed = (item: {
        [name: string]: boolean;
    }): boolean => {
        return this.requiredAttributes.every((key) => item[key] === true);
    };

    /**
     * Get Dataset-JSON metadata
     * @return An object with file metadata.
     */
    async getMetadata(): Promise<DatasetMetadata> {
        // If the file did not change, use the metadata obtained during initialization;
        if (!(await this.fileChanged()) && this.metadataLoaded === true) {
            return this.metadata;
        } else {
            if (this.isNdJson) {
                return this.getNdjsonMetadata();
            } else {
                return this.getJsonMetadata();
            }
        }
    }

    /**
     * Get Dataset-JSON metadata when the file is in JSON format.
     * @return An object with file metadata.
     */
    private async getJsonMetadata(): Promise<DatasetMetadata> {
        return new Promise((resolve, reject) => {
            this.metadataLoaded = false;
            // Metadata for ItemGroup
            const metadata: DatasetMetadata = {
                datasetJSONCreationDateTime: "",
                datasetJSONVersion: "",
                records: -1,
                name: "",
                label: "",
                columns: [],
                studyOID: "",
                metaDataVersionOID: "",
            };
            const parsedMetadata: ParsedAttributes = {
                datasetJSONCreationDateTime: false,
                datasetJSONVersion: false,
                dbLastModifiedDateTime: false,
                fileOID: false,
                originator: false,
                sourceSystem: false,
                itemGroupOID: false,
                isReferenceData: false,
                columns: false,
                records: false,
                name: false,
                label: false,
                studyOID: false,
                metaDataVersionOID: false,
                metaDataRef: false,
            };

            // Restart stream
            if (this.currentPosition !== 0 || this.stream?.destroyed) {
                if (!this.stream?.destroyed) {
                    this.stream?.destroy();
                }
                this.stream = fs.createReadStream(this.filePath, {
                    encoding: "utf8",
                });
            }

            this.stream
                .pipe(
                    JSONStream.parse(
                        "rows..*",
                        (data: string, nodePath: string) => {
                            return { path: nodePath, value: data };
                        }
                    )
                )
                .on("end", () => {
                    // Check if all required attributes are parsed after the file is fully loaded;
                    if (!this.checkAttributesParsed(parsedMetadata)) {
                        const notParsed = Object.keys(parsedMetadata).filter(
                            (key) =>
                                !parsedMetadata[key as MetadataAttributes] &&
                                this.requiredAttributes.includes(key)
                        );
                        reject(
                            new Error(
                                "Could not find required metadata elements " +
                                    notParsed.join(", ")
                            )
                        );
                    }
                    this.metadataLoaded = true;
                    this.metadata = metadata;
                    resolve(metadata);
                })
                .on("header", (data: DatasetMetadata) => {
                    // In correctly formed Dataset-JSON, all metadata attributes are present before rows
                    Object.keys(data).forEach((key) => {
                        if (Object.keys(parsedMetadata).includes(key)) {
                            (metadata as any)[key as MetadataAttributes] =
                                data[key as MetadataAttributes];
                            parsedMetadata[key as MetadataAttributes] = true;
                        }
                    });
                    // Check if all required elements were parsed
                    if (this.checkAttributesParsed(parsedMetadata)) {
                        this.metadataLoaded = true;
                        this.metadata = metadata;
                        resolve(metadata);
                        this.stream.destroy();
                    }
                })
                .on("footer", (data: DatasetMetadata) => {
                    // If not all required metadata attributes were found before rows, check if they are present after
                    Object.keys(data).forEach((key) => {
                        if (Object.keys(parsedMetadata).includes(key)) {
                            (metadata as any)[key as MetadataAttributes] =
                                data[key as MetadataAttributes];
                            parsedMetadata[key as MetadataAttributes] = true;
                        }
                    });
                    // Check if all required elements were parsed
                    if (this.checkAttributesParsed(parsedMetadata)) {
                        this.metadataLoaded = true;
                        this.metadata = metadata;
                        resolve(metadata);
                        this.stream.destroy();
                    }
                });
        });
    }

    /**
     * Get Dataset-JSON metadata when the file is in NDJSON format.
     * @return An object with file metadata.
     */
    private async getNdjsonMetadata(): Promise<DatasetMetadata> {
        return new Promise((resolve, reject) => {
            this.metadataLoaded = false;
            // All metadata is stored in the first line of the file
            let metadata: DatasetMetadata = {
                datasetJSONCreationDateTime: "",
                datasetJSONVersion: "",
                records: -1,
                name: "",
                label: "",
                columns: [],
                studyOID: "",
                metaDataVersionOID: "",
            };
            let parsedMetadata: ParsedAttributes = {
                datasetJSONCreationDateTime: false,
                datasetJSONVersion: false,
                dbLastModifiedDateTime: false,
                fileOID: false,
                originator: false,
                sourceSystem: false,
                itemGroupOID: false,
                isReferenceData: false,
                columns: false,
                records: false,
                name: false,
                label: false,
                studyOID: false,
                metaDataVersionOID: false,
                metaDataRef: false,
            };

            // Restart stream
            if (this.currentPosition !== 0 || this.stream?.destroyed) {
                if (!this.stream?.destroyed) {
                    this.stream?.destroy();
                }
                this.stream = fs.createReadStream(this.filePath, {
                    encoding: "utf8",
                });
            }

            this.rlStream = readline.createInterface({
                input: this.stream,
                crlfDelay: Infinity,
            });

            this.rlStream.on("line", (line) => {
                const data = JSON.parse(line);
                // Fill metadata with parsed attributes
                Object.keys(data).forEach((key) => {
                    if (Object.keys(parsedMetadata).includes(key)) {
                        (metadata as any)[key as MetadataAttributes] =
                            data[key as MetadataAttributes];
                        parsedMetadata[key as MetadataAttributes] = true;
                    }
                });
                // Check if all required elements were parsed
                if (this.checkAttributesParsed(parsedMetadata)) {
                    this.metadataLoaded = true;
                    this.metadata = metadata;
                    resolve(metadata);
                } else {
                    const notParsed = Object.keys(parsedMetadata).filter(
                        (key) =>
                            !parsedMetadata[key as MetadataAttributes] &&
                            this.requiredAttributes.includes(key)
                    );
                    reject(
                        new Error(
                            "Could not find required metadata elements: " +
                                notParsed.join(", ")
                        )
                    );
                }
                if (this.rlStream !== undefined) {
                    this.rlStream.close();
                }
                this.stream.destroy();
            });
        });
    }

    /**
     * Read observations.
     * @param start - The first row number to read.
     * @param length - The number of records to read.
     * @param type - The type of the returned object.
     * @param filterColumns - The list of columns to return when type is object. If empty, all columns are returned.
     * @return An array of observations.
     */
    async getData(props: {
        start: number;
        length?: number;
        type?: DataType;
        filterColumns?: string[];
    }): Promise<(ItemDataArray | ItemDataObject)[]> {
        // Check if metadata is loaded
        if (this.metadataLoaded === false) {
            await this.getMetadata();
        }

        let { filterColumns = [] } = props;

        // Convert filterColumns to lowercase for case-insensitive comparison
        filterColumns = filterColumns.map((item) => item.toLowerCase());

        // Check if metadata is loaded
        if (
            this.metadata.columns.length === 0 ||
            this.metadata.records === -1
        ) {
            return Promise.reject(
                new Error("Metadata is not loaded or there are no columns")
            );
        }
        const { start, length } = props;
        // Check if start and length are valid
        if (
            (typeof length === "number" && length <= 0) ||
            start < 0 ||
            start > this.metadata.records
        ) {
            return Promise.reject(
                new Error("Invalid start/length parameter values")
            );
        }
        if (this.isNdJson) {
            return this.getNdjsonData({ ...props, filterColumns });
        } else {
            return this.getJsonData({ ...props, filterColumns });
        }
    }

    private async getJsonData(props: {
        start: number;
        length?: number;
        type?: DataType;
        filterColumns?: string[];
    }): Promise<(ItemDataArray | ItemDataObject)[]> {

        // Default type to array;
        const { start, length, type = "array" } = props;

        const filterColumns = props.filterColumns as string[];

        return new Promise((resolve, reject) => {
            // Validate parameters
            const columnNames: string[] = [];
            if (type === "object") {
                columnNames.push(
                    ...this.metadata.columns.map((item) => item.name)
                );
            }
            // If possible, continue reading existing stream, otherwise recreate it.
            let currentPosition = this.currentPosition;
            if (this.stream.destroyed || currentPosition > start) {
                if (!this.stream.destroyed) {
                    this.stream.destroy();
                }
                this.stream = fs.createReadStream(this.filePath, {
                    encoding: "utf8",
                });
                currentPosition = 0;
                this.parser = JSONStream.parse(
                    ["rows", true],
                    (data: string, nodePath: string) => {
                        return { path: nodePath, value: data };
                    }
                ) as fs.ReadStream;
                this.stream.pipe(this.parser as unknown as fs.WriteStream);
            }

            if (this.parser === undefined) {
                reject(new Error("Could not create JSON parser"));
                return;
            }

            const currentData: (ItemDataArray | ItemDataObject)[] = [];

            this.parser
                .on("end", () => {
                    resolve(currentData);
                    this.allRowsRead = true;
                })
                .on("data", (data: { path: string; value: ItemDataArray }) => {
                    currentPosition += 1;
                    if (
                        length === undefined ||
                        (currentPosition > start &&
                            currentPosition <= start + length)
                    ) {
                        if (type === "array") {
                            currentData.push(data.value as ItemDataArray);
                        } else if (type === "object") {
                            const obj: ItemDataObject = {};
                            if (filterColumns.length === 0) {
                                columnNames.forEach((name, index) => {
                                    obj[name] = data.value[index];
                                });
                            } else {
                                // Keep only attributes specified in filterColumns
                                columnNames.forEach((name, index) => {
                                    if (
                                        filterColumns.includes(
                                            name.toLowerCase()
                                        )
                                    ) {
                                        obj[name] = data.value[index];
                                    }
                                });
                            }
                            currentData.push(obj);
                        }
                    }

                    if (
                        length !== undefined &&
                        currentPosition === start + length
                    ) {
                        const parser = this.parser as fs.ReadStream;
                        // Pause the stream and remove current event listeners
                        parser.pause();
                        parser.removeAllListeners("end");
                        parser.removeAllListeners("data");
                        this.currentPosition = currentPosition;
                        resolve(currentData);
                    }
                });
            // Resume the stream if it was paused
            if ((this.parser as unknown as { paused: boolean }).paused) {
                // Remove previous data
                this.parser.resume();
            }
        });
    }

    private async getNdjsonData(props: {
        start: number;
        length?: number;
        type?: DataType;
        filterColumns?: string[];
    }): Promise<(ItemDataArray | ItemDataObject)[]> {

        return new Promise((resolve, reject) => {

            // Default type to array;
            const { start, length, type = "array" } = props;
            const filterColumns = props.filterColumns as string[];

            // If possible, continue reading existing stream, otherwise recreate it.
            let currentPosition = this.currentPosition;
            if (this.stream.destroyed || currentPosition > start) {
                if (!this.stream.destroyed) {
                    this.stream.destroy();
                }
                this.stream = fs.createReadStream(this.filePath, {
                    encoding: "utf8",
                });
                currentPosition = 0;
                this.rlStream = readline.createInterface({
                    input: this.stream,
                    crlfDelay: Infinity,
                });
            }

            if (this.rlStream === undefined) {
                reject(new Error("Could not create readline stream"));
                return;
            }

            const columnNames: string[] = [];
            if (type === "object") {
                columnNames.push(
                    ...this.metadata.columns.map((item) => item.name)
                );
            }

            const currentData: (ItemDataArray | ItemDataObject)[] = [];

            this.rlStream
            .on("line", (line) => {
                currentPosition += 1;
                if (
                    (length === undefined ||
                        (currentPosition > start &&
                            currentPosition <= start + length)) &&
                    line.length > 0
                ) {
                    const data = JSON.parse(line);
                    if (type === "array") {
                        currentData.push(data as ItemDataArray);
                    } else if (type === "object") {
                        const obj: ItemDataObject = {};
                        if (filterColumns.length === 0) {
                            columnNames.forEach((name, index) => {
                                obj[name] = data[index];
                            });
                        } else {
                            // Keep only attributes specified in filterColumns
                            columnNames.forEach((name, index) => {
                                if (
                                    filterColumns.includes(
                                        name.toLowerCase()
                                    )
                                ) {
                                    obj[name] = data[index];
                                }
                            });
                        }
                        currentData.push(obj);
                    }
                }
                if (
                    length !== undefined &&
                    currentPosition === start + length
                ) {
                    // When pausing readline, it does not stop immidiately and can emit extra lines,
                    // so pausing approach is not yet implemented
                    if (this.rlStream !== undefined) {
                        this.rlStream.close();
                    }
                    this.stream.destroy();
                    this.currentPosition = currentPosition;
                    resolve(currentData);
                }
            })
            .on("error", (err) => {
                reject(err);
            })
            .on("close", () => {
                resolve(currentData);
                this.allRowsRead = true;
            });
        });
    }

    /**
     * Read observations as an iterable.
     * @param start - The first row number to read.
     * @param bufferLength - The number of records to read in a chunk.
     * @param type - The type of the returned object.
     * @param filterColumns - The list of columns to return when type is object. If empty, all columns are returned.
     * @return An iterable object.
     */

    async *readRecords(props?: {
        start?: number;
        bufferLength?: number;
        type?: DataType;
        filterColumns?: string[];
    }): AsyncGenerator<ItemDataArray | ItemDataObject, void, undefined> {
        // Check if metadata is loaded
        if (this.metadataLoaded === false) {
            await this.getMetadata();
        }

        const {
            start = 0,
            bufferLength = 1000,
            type,
            filterColumns,
        } = props || {};
        let currentPosition = start;

        while (true) {
            const data = await this.getData({
                start: currentPosition,
                length: bufferLength,
                type,
                filterColumns,
            });
            yield* data;

            if (this.allRowsRead === true || data.length === 0) {
                break;
            }
            currentPosition = this.currentPosition;
        }
    }

    /**
     * Get unique values observations.
     * @param columns - The list of variables for which to obtain the unique observations.
     * @param limit - The maximum number of values to store. 0 - no limit.
     * @param bufferLength - The number of records to read in a chunk.
     * @param sort - Controls whether to sort the unique values.
     * @return An array of observations.
     */
    async getUniqueValues(props: {
        columns: string[];
        limit?: number;
        bufferLength?: number;
        sort?: boolean;
    }): Promise<UniqueValues> {
        const { limit = 100, bufferLength = 1000, sort = true } = props;
        let { columns } = props;
        let result: UniqueValues = {};

        // Check if metadata is loaded
        if (this.metadataLoaded === false) {
            await this.getMetadata();
        }

        const notFoundColumns: string[] = [];
        // Use the case of the columns as specified in the metadata
        columns = columns.map((item) => {
            const column = this.metadata.columns.find(
                (column) => column.name.toLowerCase() === item.toLowerCase()
            );
            if (column === undefined) {
                notFoundColumns.push(item);
                return "";
            } else {
                return column.name as string;
            }
        });

        if (notFoundColumns.length > 0) {
            return Promise.reject(
                new Error(`Columns ${notFoundColumns.join(", ")} not found`)
            );
        }

        // Store number of unique values found
        const uniqueCount: { [name: string]: number } = {};
        columns.forEach((column) => {
            uniqueCount[column] = 0;
        });

        let isFinished = false;

        for await (const row of this.readRecords({
            bufferLength,
            type: "object",
            filterColumns: columns,
        }) as AsyncGenerator<ItemDataObject>) {
            columns.forEach((column) => {
                if (result[column] === undefined) {
                    result[column] = [];
                }
                if (
                    uniqueCount[column] < limit &&
                    row[column] !== null &&
                    !result[column].includes(row[column])
                ) {
                    result[column].push(row[column]);
                    uniqueCount[column] += 1;
                }
            });

            // Check if all unique values are found
            isFinished = Object.keys(uniqueCount).every(
                (key) => uniqueCount[key] >= limit
            );

            if (isFinished) {
                break;
            }
        }

        // Sort result
        if (sort) {
            Object.keys(result).forEach((key) => {
                result[key].sort();
            });
        }

        return result;
    }
}

export default DatasetJson;
