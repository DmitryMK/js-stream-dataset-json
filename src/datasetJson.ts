import fs from "fs";
import {
  ItemDataArray,
  ItemDescription,
  ItemGroupData,
  Data,
} from "interfaces/datasetJson.d";
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import JSONStream from "JSONStream";

type ItemGroupMetadata = Omit<ItemGroupData, "itemData">;
type DataMetadata = Omit<Data, "itemGroupData">;

// Main class for dataset JSON;
class DatasetJson {
  // Path to the file;
  filePath: string;

  // Item Group metadata;
  itemGroupMetadata: ItemGroupMetadata;
  // Item Group metadata;
  dataMetadata: DataMetadata;

  // Current position in the file;
  currentPointer: number;

  constructor(filePath: string) {
    this.filePath = filePath;
    this.currentPointer = 0;
    this.itemGroupMetadata = {
      records: -1,
      name: "",
      label: "",
      items: [],
    };

    this.dataMetadata = {
      studyOID: "",
      metaDataVersionOID: "",
    };

    // Check if file exists;
    if (!fs.existsSync(filePath)) {
      throw new Error(`Could not read file ${filePath}`);
    }
  }

  async init() {
    const metadata = await this.getMetadata();
    this.itemGroupMetadata = metadata.itemGroupMetadata;
    this.dataMetadata = metadata.dataMetadata;
  }

  async getMetadata(): Promise<{
    itemGroupMetadata: ItemGroupMetadata;
    dataMetadata: DataMetadata;
  }> {
    return new Promise((resolve, reject) => {
      const itemGroupMetadata: ItemGroupMetadata = {
        records: -1,
        name: "",
        label: "",
        items: [],
      };
      const dataMetadata: DataMetadata = {
        studyOID: "",
        metaDataVersionOID: "",
      };
      const parsedItemGroupMetadata = {
        records: false,
        name: false,
        label: false,
        items: false,
      };
      const parsedDataMetadata = {
        studyOID: false,
        metaDataVersionOID: false,
        metaDataRef: false,
      };

      const checkAttributesParsed = (
        item: { [name: string]: boolean },
        optionalAttributes: string[] = []
      ): boolean => {
        return Object.keys(item)
          .filter((key) => !optionalAttributes.includes(key))
          .every((key) => item[key] === true);
      };

      const stream = fs.createReadStream(this.filePath, {
        encoding: "utf8",
      });
      stream
        .pipe(
          JSONStream.parse(
            [
              { recurse: true },
              /name|label|items|records|metaDataVersionOID|studyOID|metaDataRef|fileOID/,
            ],
            (data: string, nodePath: string) => {
              return { path: nodePath, value: data };
            }
          )
        )
        .on("end", () => {
          if (
            !checkAttributesParsed(parsedItemGroupMetadata) ||
            !checkAttributesParsed(parsedDataMetadata, ["metaDataRef"])
          ) {
            reject(new Error("Could not find required metadata elements"));
          }
          resolve({ itemGroupMetadata, dataMetadata });
        })
        .on(
          "data",
          (data: {
            path: string;
            value: string | number | Array<ItemDescription>;
          }) => {
            const key = data.path[data.path.length - 1];

            if (Object.keys(parsedItemGroupMetadata).includes(key)) {
              itemGroupMetadata[key as "name"] = data.value as string;
              parsedItemGroupMetadata[key as keyof ItemGroupMetadata] = true;
            }

            if (Object.keys(parsedDataMetadata).includes(key)) {
              dataMetadata[key as keyof DataMetadata] = data.value as string;
              parsedDataMetadata[key as keyof DataMetadata] = true;
            }

            // Check if all required elements were parsed
            if (
              checkAttributesParsed(parsedItemGroupMetadata) &&
              checkAttributesParsed(parsedDataMetadata, ["metaDataRef"])
            ) {
              resolve({ itemGroupMetadata, dataMetadata });
              stream.destroy();
            }
          }
        );
    });
  }

  async getData(start: number, length: number): Promise<ItemDataArray[]> {
    return new Promise((resolve) => {
      const stream = fs.createReadStream(this.filePath, {
        encoding: "utf8",
      });
      const currentData: ItemDataArray[] = [];
      let currentPosition = 0;
      stream
        .pipe(
          JSONStream.parse(
            [/clinicalData|referenceData/, true, true, /itemData/, true],
            (data: string, nodePath: string) => {
              return { path: nodePath, value: data };
            }
          )
        )
        .on("end", () => {
          resolve(currentData);
        })
        .on("data", (data: { path: string; value: ItemDataArray }) => {
          currentPosition += 1;
          if (currentPosition >= start && currentPosition < start + length) {
            currentData.push(data.value);
          } else if (currentPosition === start + length + 1) {
            resolve(currentData);
            stream.destroy();
          }
        });
    });
  }
}

export default DatasetJson;
