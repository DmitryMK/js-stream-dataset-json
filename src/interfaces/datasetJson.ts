// Variable Types
export type ItemType =
    | 'string'
    | 'integer'
    | 'float'
    | 'double'
    | 'decimal'
    | 'boolean'
    | 'date'
    | 'time'
    | 'datetime'
    | 'URI'
    ;

// Target Variable Types
export type ItemTargetType =
    | 'integer'
    | 'decimal';

// Type of returned Object from getData
export type ItemDataObject = { [name: string]: string | number | boolean | null };

// The first item in the data array needs to be a number (itemGroupDataSeq)
export type ItemDataArray = Array<string | number | boolean | null >;

// Source System interface
export interface SourceSystem {
    // Name of the Source System
    name: string;
    // Version of the Source System
    version: string;
}


// Definition for Variable in the Dataset
export interface ItemDescription {
    // Unique identifier for Variable. Must correspond to ItemDef/@OID in Define-XML.
    itemOID: string;
    // Name for Variable
    name: string;
    // Label for Variable
    label: string;
    // Data type for Variable
    dataType: ItemType;
    // Indicates the data type into which the receiving system must transform the associated Dataset-JSON variable.
    targetDataType?: ItemType;
    // Length for Variable
    length?: number;
    // Display format supports data visualization of numeric float and date values.
    displayFormat?: string;
    // Indicates that this item is a key variable in the dataset structure. It also provides an ordering for the keys.
    keySequence?: number;
}


// Definition for Dataset-JSON
export interface Dataset {
    // Time of creation of the file containing the document.
    datasetJSONCreationDateTime: string;
    // Version of Dataset-JSON standard.
    datasetJSONVersion: string;
    // The total number of records in a dataset
    records: number;
    // The human readable name of the dataset
    name: string;
    // A short description of the dataset
    label: string;
    // Variable metadata
    columns: Array<ItemDescription>;
    // Data
    rows: Array<ItemDataArray>;
    // The date/time source database was last modified.
    dbLastModifiedDateTime? : string;
    // A unique identifier for this file.
    fileOID?: string;
    // The organization that generated the Dataset-JSON file.
    originator?: string;
    // The computer system or database management system that is the source of the information in this file.
    sourceSystem?: SourceSystem;
    // Unique identifier for Study. See ODM definition for study OID (ODM/Study/@OID).
    studyOID?: string;
    // Metadata for the data contained in the file.
    metaDataVersionOID?: string;
    // URL for a metadata file the describing the data.
    metaDataRef?: string;
    // Foreign key to ItemGroupDef.OID in Define / MDR
    itemGroupOID?: string;
}

// Metadata for the dataset
export type DatasetMetadata = Omit<Dataset, 'rows'>;
// Metadata Attributes
export type MetadataAttributes = keyof DatasetMetadata;
// Interface for checking which attributes are parsed
export type ParsedAttributes = {
    [name in MetadataAttributes]: boolean;
}
// Type of the object returned
export type DataType = 'array' | 'object';
export interface UniqueValues {
    [name: string]: (string | number | boolean | null)[];
}
