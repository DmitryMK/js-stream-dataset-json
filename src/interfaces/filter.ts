export type StringOperator = 'in' | 'notin' | 'eq' | 'ne' | 'starts' | 'ends' | 'contains' | 'notcontains' | 'regex';
export type DateOperator = 'lt' | 'le' | 'gt' | 'ge' | 'in' | 'notin' | 'eq' | 'ne' |
    'starts' | 'ends' | 'contains' | 'notcontains' | 'regex';
export type NumberOperator = 'lt' | 'le' | 'gt' | 'ge' | 'in' | 'notin' | 'eq' | 'ne';
export type BooleanOperator = 'eq' | 'ne';

export type Connector = 'and' | 'or';

export interface FilterCondition {
    variable: string;
    operator: StringOperator | DateOperator | NumberOperator | BooleanOperator;
    value: string | number | boolean | null | string[] | number[];
}

export interface Filter {
    conditions: FilterCondition[];
    connectors: Connector[];
    options?: {
        caseInsensitive: boolean;
    };
}

export interface ParsedFilter extends Filter {
    variableIndeces: number[];
    variableTypes: string[];
    onlyAndConnectors: boolean;
    onlyOrConnectors: boolean;
}