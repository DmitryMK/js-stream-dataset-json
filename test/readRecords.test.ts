import DatasetJson from "../src/index";
import { ItemDataObject } from "../src/interfaces/datasetJson";


test("Read the first and the last record dataset", async () => {
    const filePath = "test/data/adsl.json";

    const data = new DatasetJson(filePath);

    const result: ItemDataObject[] = [];
    let rowsCount = 0;
    for await (const row of data.readRecords({type: "object"}) as AsyncIterable<ItemDataObject>) {
        rowsCount++;

        if (rowsCount === 1 || rowsCount === data.metadata.records) {
            result.push(row);
        }
    }

    expect(result).toMatchSnapshot();
});