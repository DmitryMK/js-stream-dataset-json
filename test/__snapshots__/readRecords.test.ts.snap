// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Read the first and the last record dataset 1`] = `
[
  {
    "AGE": 63,
    "AGEGR1": "<65",
    "AGEGR1N": 1,
    "AGEU": "YEARS",
    "ARM": "Placebo",
    "AVGDD": 0,
    "BMIBL": 25.1,
    "BMIBLGR1": "25-<30",
    "COMP16FL": "Y",
    "COMP24FL": "Y",
    "COMP8FL": "Y",
    "CUMDOSE": 0,
    "DCDECOD": "COMPLETED",
    "DCSREAS": "",
    "DISCONFL": "",
    "DISONSDT": "2010-04-30",
    "DSRAEFL": "",
    "DTHFL": "",
    "DURDIS": 43.9,
    "DURDSGR1": ">=12",
    "EDUCLVL": 16,
    "EFFFL": "Y",
    "EOSSTT": "COMPLETED",
    "ETHNIC": "HISPANIC OR LATINO",
    "HEIGHTBL": 147.3,
    "ITEMGROUPDATASEQ": 1,
    "ITTFL": "Y",
    "MMSETOT": 23,
    "RACE": "WHITE",
    "RACEN": 1,
    "RFENDT": "2014-07-02",
    "RFENDTC": "2014-07-02",
    "RFSTDTC": "2014-01-02",
    "SAFFL": "Y",
    "SEX": "F",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1015",
    "TRT01A": "Placebo",
    "TRT01AN": 0,
    "TRT01P": "Placebo",
    "TRT01PN": 0,
    "TRTDURD": 182,
    "TRTEDT": "2014-07-02",
    "TRTSDT": "2014-01-02",
    "USUBJID": "01-701-1015",
    "VISIT1DT": "2013-12-26",
    "VISNUMEN": 12,
    "WEIGHTBL": 54.4,
  },
  {
    "AGE": 74,
    "AGEGR1": "65-80",
    "AGEGR1N": 2,
    "AGEU": "YEARS",
    "ARM": "Xanomeline High Dose",
    "AVGDD": 73.9,
    "BMIBL": 20.5,
    "BMIBLGR1": "<25",
    "COMP16FL": "N",
    "COMP24FL": "N",
    "COMP8FL": "Y",
    "CUMDOSE": 4212,
    "DCDECOD": "LACK OF EFFICACY",
    "DCSREAS": "Lack of Efficacy",
    "DISCONFL": "Y",
    "DISONSDT": "2010-05-03",
    "DSRAEFL": "",
    "DTHFL": "",
    "DURDIS": 31.4,
    "DURDSGR1": ">=12",
    "EDUCLVL": 16,
    "EFFFL": "Y",
    "EOSSTT": "DISCONTINUED",
    "ETHNIC": "NOT HISPANIC OR LATINO",
    "HEIGHTBL": 158.8,
    "ITEMGROUPDATASEQ": 254,
    "ITTFL": "Y",
    "MMSETOT": 19,
    "RACE": "BLACK OR AFRICAN AMERICAN",
    "RACEN": 2,
    "RFENDT": "2013-02-18",
    "RFENDTC": "2013-02-18",
    "RFSTDTC": "2012-12-17",
    "SAFFL": "Y",
    "SEX": "F",
    "SITEGR1": "718",
    "SITEID": "718",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1427",
    "TRT01A": "Xanomeline High Dose",
    "TRT01AN": 81,
    "TRT01P": "Xanomeline High Dose",
    "TRT01PN": 81,
    "TRTDURD": 57,
    "TRTEDT": "2013-02-11",
    "TRTSDT": "2012-12-17",
    "USUBJID": "01-718-1427",
    "VISIT1DT": "2012-12-13",
    "VISNUMEN": 8,
    "WEIGHTBL": 51.7,
  },
]
`;
